cpp/src/arrow/record_batch.cc

```diff
@@ -102,8 +103,10 @@ class SimpleRecordBatch : public RecordBatch {
   std::shared_ptr<Array> column(int i) const override {
     std::shared_ptr<Array> result = std::atomic_load(&boxed_columns_[i]);
     if (!result) {
-      result = MakeArray(columns_[i]);
-      std::atomic_store(&boxed_columns_[i], result);
+      auto new_array = MakeArray(columns_[i]);
+      if (std::atomic_compare_exchange_strong(&boxed_columns_[i], &result, new_array)) {
```

Comment:
```suggestion
      // Be careful not to overwrite existing entry if another thread has been calling
      // `column(i)` at the same time, since the `boxed_columns_` contents are exposed
      // by `columns()` (see GH-45371).
      if (std::atomic_compare_exchange_strong(&boxed_columns_[i], &result, new_array)) {
```